### ISSUE TITLE
FloatDomainError

### ISSUE THREAD

**Issue by alejandro-falkowski-gelato** [piotrmurach/rspec-benchmark#13]
### Are you in the right place?
* For issues or feature requests file a GitHub issue in this repository
* For general questions or discussion post on StackOverflow

### Describe the problem
NaN (FloatDomainError)

### Steps to reproduce the problem
This is hard to reproduce as I have no idea why it just happens for specific code. Though the error is happening in `lib/rspec/benchmark/iteration_matcher.rb:77:in '%'`

### Actual behaviour
NaN (FloatDomainError) is raised

### Expected behaviour
Work 😄 

### Describe your environment

* OS version: Darwin 18.7.0 Darwin Kernel Version 18.7.0: Tue Aug 20 16:57:14 PDT 2019; root:xnu-4903.271.2~2/RELEASE_X86_64
* Ruby version: ruby 2.6.4p104 (2019-08-28 revision 67798) [x86_64-darwin18]
* RSpec version: rspec (3.8.0)
* RSpec::Benchmark version: rspec (3.8.0)

**Comment by alejandro-falkowski-gelato** [comment]
When I add extra calls `.within(0.4).warmup(0.2)` seems to behave.

**Comment by ldlamarc** [comment]
I have the same error.

Quickly monkey patched it in my tests to handle the error without really resolving the underlying issue (I guess @average is 0). Code for anyone interested:

module RSpec
  module Benchmark
    module IterationMatcher
      class Matcher
        # Monkey patch to stop standard deviation NaN error
        def actual
          begin
            "%d (± %d%%) i/s" % [@average, (@stddev / @average.to_f) * 100]
          rescue FloatDomainError
            "%d (± NaN) i/s" % [@average]
          end
        end
      end
    end
  end
end

**Comment by piotrmurach** [comment]
@ldlamarc Thanks for sharing your patch. I gather the `@average` is not zero, otherwise your patch would inform `0 i/s`. Are you able to show me the spec that tends to raise this error? I'm now really keen to investigate this as it shouldn't be mathematically possible to get `0` unless somehow float doesn't have enough precision.

**Comment by piotrmurach** [comment]
@ldlamarc The `IterationMatcher` is built on top of [benchmark-perf](https://github.com/piotrmurach/benchmark-perf). Particularly, I would interested to see some example values for https://github.com/piotrmurach/benchmark-perf/blob/master/lib/benchmark/perf/iteration.rb#L90

**Comment by albanpeignier** [comment]
Same problem when using perform_at_least.

This spec works:

`expect {
        'x' * 1024 * 10
      }.to perform_under(6).ms and perform_at_least(10_000).ips`

this one raises `FloatDomainError: NaN`:

`expect {
        'x' * 1024 * 10
      }.to perform_at_least(10_000).ips`

**Comment by albanpeignier** [comment]
The problem occurs when the `run_warmup` is too long in `Iteration#run`. The `while `loop starts with `Time.now < target` at false and `measurements` stays empty.

```ruby
def run(time: 2, warmup: 1, &work)
  target = Time.now + time
  iter = 0
  measurements = []
  cycles = run_warmup(warmup: warmup, &work)

  GC.start

  while Time.now < target
  ...
```

But default the given time is 0.2 in `IterationMatcher`. Our specs are fine when we're using within method like `perform_at_least(...).within(1.second).ips`.

**Comment by piotrmurach** [comment]
@albanpeignier Thank you for bringing this up! 🙏 

Great that you found the issue! Would you have time to submit a patch with a test demonstrating the problem? [timecop](https://github.com/travisjeffery/timecop) may help to freeze time.

**Comment by piotrmurach** [comment]
@albanpeignier Once again thanks for looking into this. I've made many [changes](https://github.com/piotrmurach/benchmark-perf/blob/master/CHANGELOG.md) to the [benchmark-perf](https://github.com/piotrmurach/benchmark-perf) gem. I also updated the other [benchmark-trend](https://github.com/piotrmurach/benchmark-trend) and [benchmark-malloc](https://github.com/piotrmurach/benchmark-malloc) gems to bring them up to date with Ruby 2.7.

Thanks for your patience! I've released `v0.6.0` of `rspec-benchmark` which fixes this issue. 🎉
